### PR TITLE
[mecmeter] Initial contribution of mecmeter binding

### PR DESCRIPTION
**Pull request by kaikreuzer**
This is a new binding for the din-rail power meter from MEC, see https://www.mec.at/produkte/#mecMeter.

Signed-off-by: Kai Kreuzer <kai@openhab.org>

### CODE REVIEW

**Comment by kaikreuzer** [comment]
@openhab/add-ons-maintainers This binding is finally ready for review. Should be a fairly simple one, not that much code altogether.

File: bundles/org.openhab.binding.mecmeter/src/main/java/org/openhab/binding/mecmeter/internal/MecMeterHandlerFactory.java
Comment (wborn) [inline bundles/org.openhab.binding.mecmeter/src/main/java/org/openhab/binding/mecmeter/internal/MecMeterHandlerFactory.java:46]: In Java 11 this can be more clear and shorter with:

```suggestion
    private static final Set<ThingTypeUID> SUPPORTED_THING_TYPES_UIDS = Set.of(THING_TYPE_METER);
```

If you move this to the `MecMeterBindingConstants`, then it can also be reused in the `MecMeterDiscoveryParticipant`.

File: bundles/org.openhab.binding.mecmeter/src/main/resources/OH-INF/thing/thing-types.xml
Comment (wborn) [inline bundles/org.openhab.binding.mecmeter/src/main/resources/OH-INF/thing/thing-types.xml:51]:
```suggestion
				<description>The IP address of the mecMeter</description>
```

File: bundles/org.openhab.binding.mecmeter/src/main/java/org/openhab/binding/mecmeter/internal/discovery/MecMeterDiscoveryParticipant.java
Comment (wborn) [inline bundles/org.openhab.binding.mecmeter/src/main/java/org/openhab/binding/mecmeter/internal/discovery/MecMeterDiscoveryParticipant.java:89]: Why use 2 if you add 3 entries? 

```suggestion
        Map<String, Object> properties = new HashMap<>(3);
```

File: bundles/org.openhab.binding.mecmeter/src/main/java/org/openhab/binding/mecmeter/internal/api/MecMeterResponse.java
Comment (wborn) [inline bundles/org.openhab.binding.mecmeter/src/main/java/org/openhab/binding/mecmeter/internal/api/MecMeterResponse.java:13]: SAT warnings about missing `@NonNullByDefault` can be prevented by moving it to another package:

```suggestion
package org.openhab.binding.mecmeter.internal.dto;
```

File: bundles/org.openhab.binding.mecmeter/src/main/java/org/openhab/binding/mecmeter/internal/api/MecMeterResponse.java
Comment (wborn) [inline bundles/org.openhab.binding.mecmeter/src/main/java/org/openhab/binding/mecmeter/internal/api/MecMeterResponse.java:213]:
```suggestion
    /* Constants */
```

File: bundles/org.openhab.binding.mecmeter/src/main/java/org/openhab/binding/mecmeter/internal/discovery/MecMeterDiscoveryParticipant.java
Comment (wborn) [inline bundles/org.openhab.binding.mecmeter/src/main/java/org/openhab/binding/mecmeter/internal/discovery/MecMeterDiscoveryParticipant.java:97]: Useless assignment:

```suggestion
        return DiscoveryResultBuilder.create(uid).withProperties(properties).withLabel(label)
                .withRepresentationProperty(Thing.PROPERTY_SERIAL_NUMBER).build();
```

File: bundles/org.openhab.binding.mecmeter/README.MD
Comment (wborn) [inline bundles/org.openhab.binding.mecmeter/README.MD:26]: The IP is not mandatory according to the configuration in thing-types.xml.

File: bundles/org.openhab.binding.mecmeter/src/main/java/org/openhab/binding/mecmeter/MecMeterDeviceConfiguration.java
Comment (wborn) [inline bundles/org.openhab.binding.mecmeter/src/main/java/org/openhab/binding/mecmeter/MecMeterDeviceConfiguration.java:74]: The password is required so it cannot be null.

File: bundles/org.openhab.binding.mecmeter/src/main/java/org/openhab/binding/mecmeter/handler/MecMeterHandler.java
Comment (wborn) [inline bundles/org.openhab.binding.mecmeter/src/main/java/org/openhab/binding/mecmeter/handler/MecMeterHandler.java:70]:
```suggestion
    private @Nullable ScheduledFuture<?> pollFuture;
```

File: bundles/org.openhab.binding.mecmeter/src/main/java/org/openhab/binding/mecmeter/handler/MecMeterHandler.java
Comment (wborn) [inline bundles/org.openhab.binding.mecmeter/src/main/java/org/openhab/binding/mecmeter/handler/MecMeterHandler.java:73]:
```suggestion
    private @Nullable MecMeterResponse powerMeterResponse;
```

File: bundles/org.openhab.binding.mecmeter/src/main/java/org/openhab/binding/mecmeter/handler/MecMeterHandler.java
Comment (wborn) [inline bundles/org.openhab.binding.mecmeter/src/main/java/org/openhab/binding/mecmeter/handler/MecMeterHandler.java:129]: There is no such parameter.

File: bundles/org.openhab.binding.mecmeter/src/main/java/org/openhab/binding/mecmeter/handler/MecMeterHandler.java
Comment (wborn) [inline bundles/org.openhab.binding.mecmeter/src/main/java/org/openhab/binding/mecmeter/handler/MecMeterHandler.java:139]:
```suggestion
     * @return MecMeterResponse class where json values "are saved"
```

File: bundles/pom.xml
Comment (wborn) [inline bundles/pom.xml:181]: It also needs to be added to the BOM pom.xml.

**Review by wborn — CHANGES_REQUESTED** [review]
Thank you for migrating the code to OH3. :+1:  I placed a few review comments for my findings below.

File: bundles/org.openhab.binding.mecmeter/src/main/java/org/openhab/binding/mecmeter/MecMeterDeviceConfiguration.java
Comment (fwolter) [reply bundles/org.openhab.binding.mecmeter/src/main/java/org/openhab/binding/mecmeter/MecMeterDeviceConfiguration.java:74]: What happens if the user forgets to specify the parameter in textual configuration?

File: bundles/org.openhab.binding.mecmeter/src/main/java/org/openhab/binding/mecmeter/MecMeterDeviceConfiguration.java
Comment (kaikreuzer) [reply bundles/org.openhab.binding.mecmeter/src/main/java/org/openhab/binding/mecmeter/MecMeterDeviceConfiguration.java:20]: Missed to add them. Already got their e-mail addresses, I will squash the commits and add them to the sign-off.

File: bundles/org.openhab.binding.mecmeter/src/main/java/org/openhab/binding/mecmeter/MecMeterDeviceConfiguration.java
Comment (kaikreuzer) [reply bundles/org.openhab.binding.mecmeter/src/main/java/org/openhab/binding/mecmeter/MecMeterDeviceConfiguration.java:22]: done.

File: bundles/org.openhab.binding.mecmeter/src/main/java/org/openhab/binding/mecmeter/MecMeterDeviceConfiguration.java
Comment (kaikreuzer) [reply bundles/org.openhab.binding.mecmeter/src/main/java/org/openhab/binding/mecmeter/MecMeterDeviceConfiguration.java:29]: done.

File: bundles/org.openhab.binding.mecmeter/src/main/java/org/openhab/binding/mecmeter/MecMeterDeviceConfiguration.java
Comment (kaikreuzer) [reply bundles/org.openhab.binding.mecmeter/src/main/java/org/openhab/binding/mecmeter/MecMeterDeviceConfiguration.java:45]: removed.

File: bundles/org.openhab.binding.mecmeter/src/main/resources/OH-INF/thing/thing-types.xml
Comment (kaikreuzer) [reply bundles/org.openhab.binding.mecmeter/src/main/resources/OH-INF/thing/thing-types.xml:52]: Actually it is a required value, I have changed the "required" field accordingly.

File: bundles/org.openhab.binding.mecmeter/src/main/resources/OH-INF/thing/thing-types.xml
Comment (kaikreuzer) [reply bundles/org.openhab.binding.mecmeter/src/main/resources/OH-INF/thing/thing-types.xml:43]: removed

File: bundles/org.openhab.binding.mecmeter/src/main/java/org/openhab/binding/mecmeter/internal/discovery/MecMeterDiscoveryParticipant.java
Comment (kaikreuzer) [reply bundles/org.openhab.binding.mecmeter/src/main/java/org/openhab/binding/mecmeter/internal/discovery/MecMeterDiscoveryParticipant.java:103]: Looks like some old testing code. Removed it.

File: bundles/org.openhab.binding.mecmeter/src/main/java/org/openhab/binding/mecmeter/MecMeterDeviceConfiguration.java
Comment (kaikreuzer) [reply bundles/org.openhab.binding.mecmeter/src/main/java/org/openhab/binding/mecmeter/MecMeterDeviceConfiguration.java:68]: Indeed. As the IP is mandatory, I only left the blank check.

File: bundles/org.openhab.binding.mecmeter/src/main/java/org/openhab/binding/mecmeter/internal/MecMeterHandlerFactory.java
Comment (kaikreuzer) [reply bundles/org.openhab.binding.mecmeter/src/main/java/org/openhab/binding/mecmeter/internal/MecMeterHandlerFactory.java:46]: done.

File: bundles/org.openhab.binding.mecmeter/src/main/resources/OH-INF/thing/thing-types.xml
Comment (kaikreuzer) [reply bundles/org.openhab.binding.mecmeter/src/main/resources/OH-INF/thing/thing-types.xml:46]: fixed

File: bundles/org.openhab.binding.mecmeter/src/main/java/org/openhab/binding/mecmeter/internal/discovery/MecMeterDiscoveryParticipant.java
Comment (kaikreuzer) [reply bundles/org.openhab.binding.mecmeter/src/main/java/org/openhab/binding/mecmeter/internal/discovery/MecMeterDiscoveryParticipant.java:89]: fixed.

File: bundles/org.openhab.binding.mecmeter/src/main/java/org/openhab/binding/mecmeter/internal/api/MecMeterResponse.java
Comment (kaikreuzer) [reply bundles/org.openhab.binding.mecmeter/src/main/java/org/openhab/binding/mecmeter/internal/api/MecMeterResponse.java:13]: Didn't know, thanks. Moved.

File: bundles/org.openhab.binding.mecmeter/src/main/java/org/openhab/binding/mecmeter/internal/api/MecMeterResponse.java
Comment (kaikreuzer) [reply bundles/org.openhab.binding.mecmeter/src/main/java/org/openhab/binding/mecmeter/internal/api/MecMeterResponse.java:213]: fixed.

File: bundles/org.openhab.binding.mecmeter/src/main/java/org/openhab/binding/mecmeter/internal/discovery/MecMeterDiscoveryParticipant.java
Comment (kaikreuzer) [reply bundles/org.openhab.binding.mecmeter/src/main/java/org/openhab/binding/mecmeter/internal/discovery/MecMeterDiscoveryParticipant.java:97]: fixed.

File: bundles/org.openhab.binding.mecmeter/README.MD
Comment (kaikreuzer) [reply bundles/org.openhab.binding.mecmeter/README.MD:26]: Changed in the thing-types.xml. It is indeed mandatory.

File: bundles/org.openhab.binding.mecmeter/src/main/java/org/openhab/binding/mecmeter/MecMeterDeviceConfiguration.java
Comment (kaikreuzer) [reply bundles/org.openhab.binding.mecmeter/src/main/java/org/openhab/binding/mecmeter/MecMeterDeviceConfiguration.java]: @fwolter The Thing won't be initialized, if some mandatory parameter is missing.

File: bundles/org.openhab.binding.mecmeter/src/main/java/org/openhab/binding/mecmeter/handler/MecMeterHandler.java
Comment (kaikreuzer) [reply bundles/org.openhab.binding.mecmeter/src/main/java/org/openhab/binding/mecmeter/handler/MecMeterHandler.java:72]: fixed.

File: bundles/org.openhab.binding.mecmeter/src/main/java/org/openhab/binding/mecmeter/handler/MecMeterHandler.java
Comment (kaikreuzer) [reply bundles/org.openhab.binding.mecmeter/src/main/java/org/openhab/binding/mecmeter/handler/MecMeterHandler.java:75]: fixed.

File: bundles/org.openhab.binding.mecmeter/src/main/java/org/openhab/binding/mecmeter/handler/MecMeterHandler.java
Comment (kaikreuzer) [reply bundles/org.openhab.binding.mecmeter/src/main/java/org/openhab/binding/mecmeter/handler/MecMeterHandler.java:131]: fixed.

File: bundles/org.openhab.binding.mecmeter/src/main/java/org/openhab/binding/mecmeter/handler/MecMeterHandler.java
Comment (kaikreuzer) [reply bundles/org.openhab.binding.mecmeter/src/main/java/org/openhab/binding/mecmeter/handler/MecMeterHandler.java:140]: fixed.

File: bundles/org.openhab.binding.mecmeter/src/main/java/org/openhab/binding/mecmeter/handler/MecMeterHandler.java
Comment (kaikreuzer) [reply bundles/org.openhab.binding.mecmeter/src/main/java/org/openhab/binding/mecmeter/handler/MecMeterHandler.java]: Indeed. Fixed.

File: bundles/org.openhab.binding.mecmeter/src/main/java/org/openhab/binding/mecmeter/handler/MecMeterHandler.java
Comment (kaikreuzer) [reply bundles/org.openhab.binding.mecmeter/src/main/java/org/openhab/binding/mecmeter/handler/MecMeterHandler.java:141]: fixed.

File: bundles/org.openhab.binding.mecmeter/src/main/java/org/openhab/binding/mecmeter/handler/MecMeterHandler.java
Comment (kaikreuzer) [reply bundles/org.openhab.binding.mecmeter/src/main/java/org/openhab/binding/mecmeter/handler/MecMeterHandler.java:160]: Well, it was VERY successful ;-). Fixed.

File: bundles/org.openhab.binding.mecmeter/src/main/resources/OH-INF/thing/thing-types.xml
Comment (kaikreuzer) [reply bundles/org.openhab.binding.mecmeter/src/main/resources/OH-INF/thing/thing-types.xml:380]: fixed.

File: bundles/pom.xml
Comment (kaikreuzer) [reply bundles/pom.xml:181]: Thought I did, but you are right. Added it.

**Comment by kaikreuzer** [comment]
@wborn Thanks for your review! All comments are addressed now.

File: bundles/org.openhab.binding.mecmeter/README.MD
Comment (wborn) [inline bundles/org.openhab.binding.mecmeter/README.MD:25]: It may help to also document that this parameter is mandatory.

File: bundles/org.openhab.binding.mecmeter/src/main/java/org/openhab/binding/mecmeter/handler/MecMeterHandler.java
Comment (wborn) [inline bundles/org.openhab.binding.mecmeter/src/main/java/org/openhab/binding/mecmeter/handler/MecMeterHandler.java:170]: The error log level used in this method seems a bit too severe. The debug/error logging in this method also isn't that useful. That's because a very similar `errorMsg` is always used to update the Thing status. So it might be better to remove all the debug/error logging. WDYT?

**Review by wborn — CHANGES_REQUESTED** [review]
Thanks for quickly fixing all my review comments!

After going through the code I have 2 more comments:

File: bundles/org.openhab.binding.mecmeter/src/main/java/org/openhab/binding/mecmeter/handler/MecMeterHandler.java
Comment (kaikreuzer) [reply bundles/org.openhab.binding.mecmeter/src/main/java/org/openhab/binding/mecmeter/handler/MecMeterHandler.java:170]: No, these cases should indeed only happen if there's a bug in the binding - therefore an error is logged, so that people will report the bug.

File: bundles/org.openhab.binding.mecmeter/README.MD
Comment (kaikreuzer) [reply bundles/org.openhab.binding.mecmeter/README.MD]: done.

File: bundles/org.openhab.binding.mecmeter/src/main/java/org/openhab/binding/mecmeter/handler/MecMeterHandler.java
Comment (wborn) [reply bundles/org.openhab.binding.mecmeter/src/main/java/org/openhab/binding/mecmeter/handler/MecMeterHandler.java:170]: If they are real bugs than we need to keep them indeed. :-)

**Review by wborn — APPROVED** [review]
LGTM now! If another @openhab/add-ons-maintainers also approves the PR then we can merge this. :rocket:

**Comment by wborn** [comment]
While we wait for the second review, can you also add the binding dir to the CODEOWNERS file @kaikreuzer?

**Comment by kaikreuzer** [comment]
> can you also add the binding dir to the CODEOWNERS file 

Sure, done!

File: bundles/org.openhab.binding.mecmeter/src/main/java/org/openhab/binding/mecmeter/handler/MecMeterHandler.java
Comment (fwolter) [inline bundles/org.openhab.binding.mecmeter/src/main/java/org/openhab/binding/mecmeter/handler/MecMeterHandler.java:155]: I think this should be warn, if the remote fails.

File: bundles/org.openhab.binding.mecmeter/src/main/java/org/openhab/binding/mecmeter/handler/MecMeterHandler.java
Comment (fwolter) [inline bundles/org.openhab.binding.mecmeter/src/main/java/org/openhab/binding/mecmeter/handler/MecMeterHandler.java:403]: Are you sure these are Energy? The names sound dimensionless.

File: bundles/org.openhab.binding.mecmeter/src/main/resources/OH-INF/thing/thing-types.xml
Comment (fwolter) [inline bundles/org.openhab.binding.mecmeter/src/main/resources/OH-INF/thing/thing-types.xml:484]: This exceeds 25 chars.

**Review by fwolter — CHANGES_REQUESTED** [review]
There are two compiler warnings left:
```
[WARNING] MecMeterHandler.java:[101,13] Potential null pointer access: this expression has a '@Nullable' type
[WARNING] MecMeterHandler.java:[117,13] Potential null pointer access: this expression has a '@Nullable' type
```

File: bundles/org.openhab.binding.mecmeter/src/main/java/org/openhab/binding/mecmeter/handler/MecMeterHandler.java
Comment (kaikreuzer) [reply bundles/org.openhab.binding.mecmeter/src/main/java/org/openhab/binding/mecmeter/handler/MecMeterHandler.java:155]: https://www.openhab.org/docs/developer/guidelines.html#f-logging says:

> error logging should only be used
> - in case some code fails irrecoverably and the user should report it as a severe bug.

If we get a response, which is not a HTTP 200 here, I think we have a severe bug here, which should be reported as we seem to be doing invalid requests that make the binding disfunctional.

File: bundles/org.openhab.binding.mecmeter/src/main/resources/OH-INF/thing/thing-types.xml
Comment (kaikreuzer) [reply bundles/org.openhab.binding.mecmeter/src/main/resources/OH-INF/thing/thing-types.xml]: shortened.

File: bundles/org.openhab.binding.mecmeter/src/main/java/org/openhab/binding/mecmeter/handler/MecMeterHandler.java
Comment (kaikreuzer) [reply bundles/org.openhab.binding.mecmeter/src/main/java/org/openhab/binding/mecmeter/handler/MecMeterHandler.java]: Good point. It was in the original code with a unit, but I agree with you that it does not make sense. Have changed it to dimensionless.

**Review by fwolter — APPROVED** [review]
LGTM

**Pull request merged**
